### PR TITLE
Feature/time stamp provider monotonic

### LINKED ISSUE #634: Storage transactions time stamp not monotone increasing
MicroStream Version: 08.01.01-MS-GA 

The current default implementation of one.microstream.storage.types.StorageTimestampProvider is not monotone increasing. It depends on the system clock (System.currentTimeMillis()) which can be set to any values (e.g.: user action, NTP maybe). If set to a smaller time value then already used as time stamp in the transaction log an inconsistent timestamp error can happen at the next startup.

### PR DESCRIPTION
**Pull request by hg-ms**
- added new StorageTimestampProviders MonotonicTime and MonotonicCounter
- set MonotonicTime as default StorageTimestampProviders
- initializing StorageTimestampProviders with latest value from transactionslog

### CODE REVIEW

File: storage/storage/src/main/java/one/microstream/storage/types/StorageChannelTaskInitialize.java [hunk 1]
```diff
@@ -227,6 +240,12 @@ public synchronized StorageIdAnalysis idAnalysis()
 			);
 		}
 
+		@Override
```
Comment (fh-ms) [inline storage/storage/src/main/java/one/microstream/storage/types/StorageChannelTaskInitialize.java:246]: Maybe field name should match method name

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java [hunk 2]
```diff
@@ -30,6 +32,15 @@ public interface StorageTimestampProvider
 	 */
 	public long currentNanoTimestamp();
 	
+	/**
+	 * Set the base value used to create the time stamps
```
Comment (fh-ms) [inline storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java]: timestamp is one word

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java [hunk 3]
```diff
@@ -30,6 +32,15 @@ public interface StorageTimestampProvider
...
+	 * Implementations are allowed to ignore it.
```
Comment (fh-ms) [inline storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java:37]: timestamp is one word

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java [hunk 4]
```diff
@@ -50,6 +61,95 @@ public synchronized long currentNanoTimestamp()
...
+			
+			if(currentTimeNanos <= this.lastTimeNanos)
+			{
+				return ++this.lastTimeNanos;
+			}
+			return this.lastTimeNanos = currentTimeNanos;
+		}
+		
+		/**
+		 * Set to new base value only if the new value is lager then the current one.
```
Comment (fh-ms) [inline storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java]: lager -> larger

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java [hunk 5]
```diff
@@ -50,6 +61,95 @@ public synchronized long currentNanoTimestamp()
...
+			{
+				this.lastTimeNanos = timeNs;
+			}
+			return this.lastTimeNanos;
+		}
+	}
+	
+	/**
+	 * An implementation of {@link StorageTimestampProvider} that provides an strictly monotonic increasing
+	 * long value instead of a time value. This implementation does not relay on any time based value
```
Comment (fh-ms) [inline storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java]: relay -> rely

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java [hunk 6]
```diff
@@ -50,6 +61,95 @@ public synchronized long currentNanoTimestamp()
...
+		 * @return a strictly monotone increasing long value.
+		 */
+		@Override
+		public long currentNanoTimestamp()
+		{
+			return this.lastValue.incrementAndGet();
+		}
+
+		/**
+		 * Set to new base value only if the new value is lager then the current one.
```
Comment (fh-ms) [inline storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java]: lager -> larger

**Review by fh-ms — CHANGES_REQUESTED** [review]

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java (hunk 2 above)
Comment (hg-ms) [reply storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java]: ok, done

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java (hunk 3 above)
Comment (hg-ms) [reply storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java:37]: ok, done

File: storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java (hunk 4 above)
Comment (hg-ms) [reply storage/storage/src/main/java/one/microstream/storage/types/StorageTimestampProvider.java]: ok, done

**Review by fh-ms — APPROVED** [review]

**Pull request merged**